### PR TITLE
Features/region proposal network losses small fixes

### PR DESCRIPTION
**Pull request by hgaiser**
This PR fixes some small issues that caused the model not to compile.

In addition I noticed that all inputs are shaped including their batch dimension. I wasn't completely aware of this and I'm not sure the current implementation handles it correctly (ie. it might assume a shape of (None, 4) while it is actually (None, None, 4)). @jhung0 @0x00b1 are you aware of this?

### CODE REVIEW

**Comment by jhung0** [comment]
Oh does it compile now? 
Yeah, I noticed that issue, and we should be consistent about it...so let's try to make sure batch dimension is included

File: keras_rcnn/models.py [hunk 1]
```diff
@@ -30,15 +30,15 @@ def __init__(self, inputs, features, heads, rois):
 
         rpn_prediction = keras.layers.concatenate([rpn_classification, rpn_regression])
 
-        proposals = keras_rcnn.layers.object_detection.ObjectProposal(rois)([im_info, rpn_regression, rpn_classification])
-        anchors = keras_rcnn.layers.object_detection.ProposalTarget()([rpn_classification, gt_boxes, im_info])
+        proposals                        = keras_rcnn.layers.object_detection.ObjectProposal(rois)([im_info, rpn_regression, rpn_classification])
```
Comment (jhung0) [inline keras_rcnn/models.py]: Remove the `object_detection` part of this and the next line

**Comment by hgaiser** [comment]
Yeah with these changes I got a successful model summary (not really the same as compiling I guess).

File: keras_rcnn/models.py (hunk 1 above)
Comment (hgaiser) [inline keras_rcnn/models.py]: Done.

File: keras_rcnn/layers/object_detection/_proposal_target.py [hunk 2]
```diff
@@ -52,7 +52,13 @@ def call(self, inputs, **kwargs):
...
-        return labels, bbox_reg_targets
+        #TODO: Why is bbox_reg_targets' shape (5, ?, 4)? Why is gt_boxes' shape (None, None, 4) and not (None, 4)?
+        bbox_reg_targets = keras.backend.reshape(bbox_reg_targets, (-1, 4))
+        return [labels, bbox_reg_targets]
 
     def compute_output_shape(self, input_shape):
-        return (None,), (None, 4)
+        return [(None,), (None, 4)]
+
+    def compute_mask(self, inputs, mask=None):
```
Comment (jhung0) [inline keras_rcnn/layers/object_detection/_proposal_target.py]: why is it required?

File: keras_rcnn/models.py [hunk 3]
```diff
@@ -30,15 +30,15 @@ def __init__(self, inputs, features, heads, rois):
 
         rpn_prediction = keras.layers.concatenate([rpn_classification, rpn_regression])
 
-        proposals = keras_rcnn.layers.object_detection.ObjectProposal(rois)([im_info, rpn_regression, rpn_classification])
-        anchors = keras_rcnn.layers.object_detection.ProposalTarget()([rpn_classification, gt_boxes, im_info])
+        proposals                        = keras_rcnn.layers.ObjectProposal(rois)([im_info, rpn_regression, rpn_classification])
```
Comment (jhung0) [inline keras_rcnn/models.py]: Maybe it's not necessary, but it'd be clearer to have `maximum_proposals=rois` instead of just `rois`
Also, ObjectProposal and ProposalTarget have im_info in different orders...we should be consistent. Could you switch it so that im_info comes at the end?

**Comment by jhung0** [comment]
Are you able to train?

**Comment by jhung0** [comment]
```
# Bounding boxes are the ground truth bounding boxes
bounding_boxes = keras.layers.Input((4,))

# Metadata is a 3-tuple that contains the feature width, feature height, and scale
metadata = keras.layers.Input((3,))

a, b = ProposalTarget()([bounding_boxes, metadata])

model = keras.models.Model([bounding_boxes, metadata], [a, b])
```
isn't working

File: keras_rcnn/layers/object_detection/_proposal_target.py (hunk 2 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/_proposal_target.py]: It is trying to call `output_mask` for all outputs. I'm not sure why it isn't required for a single output, but for multiple outputs it seems to be required (https://github.com/fchollet/keras/issues/3061). Perhaps something we should try to fix in Keras.

File: keras_rcnn/models.py (hunk 3 above)
Comment (hgaiser) [inline keras_rcnn/models.py]: I agree. Will you make this change or should I?

**Comment by hgaiser** [comment]
No I didn't try training yet (are the loss layers used already?). I only got a seemingly logical model summary.
